src/coroutine.rs

```diff
@@ -0,0 +1,218 @@
+//! Python coroutine implementation, used notably when wrapping `async fn`
+//! with `#[pyfunction]`/`#[pymethods]`.
+use crate::exceptions::{PyRuntimeError, PyStopIteration};
+use crate::pyclass::IterNextOutput;
+use crate::sync::GILOnceCell;
+use crate::types::PyIterator;
+use crate::{intern, IntoPy, Py, PyAny, PyErr, PyObject, PyResult, Python};
+use futures::task::ArcWake;
+use pyo3_macros::{pyclass, pymethods};
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+const COROUTINE_REUSED_ERROR: &str = "cannot reuse already awaited coroutine";
+
+/// Python coroutine wrapping a [`Future`].
+#[pyclass(crate = "crate")]
+pub struct Coroutine {
+    future: Option<Pin<Box<dyn Future<Output = PyResult<PyObject>> + Send>>>,
+    #[allow(clippy::type_complexity)]
+    throw_callback: Option<Box<dyn FnMut(&PyAny) + Send>>,
+    waker: Option<Arc<PyFutureWaker>>,
+}
+
+impl Coroutine {
+    /// Wrap a boxed future in to a Python coroutine.
+    ///
+    /// Coroutine `send` polls the wrapped future, ignoring the value passed
+    /// (should always be `None` anyway).
+    ///
+    /// If `throw_callback` has been provided, coroutine `throw` calls it
+    /// with the exception passed, and poll the wrapped future.
+    /// Otherwise, the exception is reraised and the future dropped.
+    #[allow(clippy::type_complexity)]
+    pub fn new(
+        future: Pin<Box<dyn Future<Output = PyResult<PyObject>> + Send>>,
+        throw_callback: Option<Box<dyn FnMut(&PyAny) + Send>>,
```

Comment:
If I read correctly this is a more advanced API to do with cancellation rather than just `async fn` syntax?